cupy/cuda/memory.pyx

```diff
@@ -394,6 +412,51 @@ cdef class SingleDeviceMemoryPool:
             DeprecationWarning)
         self.free_all_blocks()
 
+    cpdef size_t realloc(self, size_t ptr, Py_ssize_t new_size):
+        """Reallocates the given area of memory."""
+        cdef Memory cur_mem
+        cdef Memory new_mem
+        cdef MemoryPointer memptr
+        cur_size = self._in_use[ptr].size
+        try:
+            new_mem = self._alloc(new_size).mem
+        except runtime.CUDARuntimeError as e:
+            if e.status != runtime.errorMemoryAllocation:
+                raise
+            runtime.deviceSynchronize()
+            new_mem = None
+        if new_mem is None:
+            return ptr
+        # memory copy
+        cur_mem = self._in_use.pop(ptr, None)
+        _size = min(cur_size, new_size)
+        runtime.memcpy(new_mem.ptr, cur_mem.ptr, _size,
+                       runtime.memcpyDeviceToDevice)
+        runtime.deviceSynchronize()
+        del cur_mem
+        # update info
+        memptr = self._in_use_memptr.pop(ptr, None)()
+        memptr.mem.ptr = new_mem.ptr
+        memptr.mem.size = new_mem.size
+        memptr.set_ptr(new_mem.ptr)
+        self._in_use[new_mem.ptr] = new_mem
+        self._in_use_memptr[new_mem.ptr] = weakref.ref(memptr)
+
+        return new_mem.ptr
+
+    cpdef realloc_all(self):
+        """Reallocates the memory of all in-use arrays"""
+        _in_use = collections.defaultdict(list)
+        for ptr in self._in_use.keys():
+            mem = self._in_use[ptr]
+            _in_use[mem.size].append(ptr)
+
+        for size in sorted(_in_use.keys()):
+            for ptr in sorted(_in_use[size]):
+                new_ptr = self.realloc(ptr, size)
+                if new_ptr > ptr:
+                    self.realloc(new_ptr, size)
```

Comment:
Could you explain why we call realloc again here?

Reply:
This is based on an assumption that logical address are almost linearly mapped to physical address. If the assumption is correct, you can make larger continuous free space in the high address area by calling the realloc() twice when new address > old address.
Note that there is no guarantee that the assumption is correct.

Reply:
Okay, thanks.